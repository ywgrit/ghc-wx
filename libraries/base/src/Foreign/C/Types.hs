-- |
--
-- Module      :  Foreign.C.Types
-- Copyright   :  (c) The FFI task force 2001
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  ffi@haskell.org
-- Stability   :  provisional
-- Portability :  portable
--
-- Mapping of C types to corresponding Haskell types.
--

module Foreign.C.Types
    (-- *  Representations of C types
     -- $ctypes
     -- **  #platform# Platform differences
     -- |  This module contains platform specific information about types.
     -- __/As such, the types presented on this page reflect the/__
     -- __/platform on which the documentation was generated and may/__
     -- __/not coincide with the types on your platform./__
     -- **  Integral types
     -- |  These types are represented as @newtype@s of
     -- types in "Data.Int" and "Data.Word", and are instances of
     -- 'Prelude.Eq', 'Prelude.Ord', 'Prelude.Num', 'Prelude.Read',
     -- 'Prelude.Show', 'Prelude.Enum', 'Data.Typeable.Typeable',
     -- 'Storable', 'Prelude.Bounded', 'Prelude.Real', 'Prelude.Integral'
     -- and 'Bits'.
     CChar(..),
     CSChar(..),
     CUChar(..),
     CShort(..),
     CUShort(..),
     CInt(..),
     CUInt(..),
     CLong(..),
     CULong(..),
     CPtrdiff(..),
     CSize(..),
     CWchar(..),
     CSigAtomic(..),
     CLLong(..),
     CULLong(..),
     CBool(..),
     CIntPtr(..),
     CUIntPtr(..),
     CIntMax(..),
     CUIntMax(..),
     -- **  Numeric types
     -- |  These types are represented as @newtype@s of basic
     -- foreign types, and are instances of
     -- 'Prelude.Eq', 'Prelude.Ord', 'Prelude.Num', 'Prelude.Read',
     -- 'Prelude.Show', 'Prelude.Enum', 'Data.Typeable.Typeable' and
     -- 'Storable'.
     CClock(..),
     CTime(..),
     CUSeconds(..),
     CSUSeconds(..),
     -- |  To convert 'CTime' to 'Data.Time.UTCTime', use the following:
     --
     -- > \t -> posixSecondsToUTCTime (realToFrac t :: POSIXTime)

     -- **  Floating types
     -- |  These types are represented as @newtype@s of
     -- 'Prelude.Float' and 'Prelude.Double', and are instances of
     -- 'Prelude.Eq', 'Prelude.Ord', 'Prelude.Num', 'Prelude.Read',
     -- 'Prelude.Show', 'Prelude.Enum', 'Data.Typeable.Typeable', 'Storable',
     -- 'Prelude.Real', 'Prelude.Fractional', 'Prelude.Floating',
     -- 'Prelude.RealFrac' and 'Prelude.RealFloat'. That does mean
     -- that `CFloat`'s (respectively `CDouble`'s) instances of
     -- 'Prelude.Eq', 'Prelude.Ord', 'Prelude.Num' and
     -- 'Prelude.Fractional' are as badly behaved as `Prelude.Float`'s
     -- (respectively `Prelude.Double`'s).
     CFloat(..),
     CDouble(..),
     -- XXX GHC doesn't support CLDouble yet
     -- , CLDouble(..)
     -- **  Other types
     CFile,
     CFpos,
     CJmpBuf
     ) where

import GHC.Internal.Foreign.C.Types
