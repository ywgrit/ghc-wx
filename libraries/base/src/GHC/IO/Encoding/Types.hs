{-# LANGUAGE Safe #-}
{-# LANGUAGE MagicHash #-}
{-# LANGUAGE PatternSynonyms #-}

-- |
--
-- Module      :  GHC.IO.Encoding.Types
-- Copyright   :  (c) The University of Glasgow, 2008-2009
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  internal
-- Portability :  non-portable
--
-- Types for text encoding/decoding
--

module GHC.IO.Encoding.Types
    (BufferCodec(..),
     TextEncoding(..),
     TextEncoder,
     TextDecoder,
     CodeBuffer,
     EncodeBuffer,
     DecodeBuffer,
     CodingProgress(..),
     DecodeBuffer#,
     EncodeBuffer#,
     DecodingBuffer#,
     EncodingBuffer#
     ) where

import GHC.Internal.IO.Encoding.Types
