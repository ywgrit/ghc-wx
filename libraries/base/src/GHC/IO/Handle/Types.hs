{-# LANGUAGE Safe #-}
{-# OPTIONS_HADDOCK not-home #-}

-- |
-- Module      :  GHC.IO.Handle.Types
-- Copyright   :  (c) The University of Glasgow, 1994-2009
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  internal
-- Portability :  non-portable
--
-- Basic types for the implementation of IO Handles.
--
-- /The API of this module is unstable and not meant to be consumed by the general public./
-- If you absolutely must depend on it, make sure to use a tight upper
-- bound, e.g., @base < 4.X@ rather than @base < 5@, because the interface can
-- change rapidly without much warning.
--

module GHC.IO.Handle.Types (
      Handle(..), Handle__(..), showHandle,
      checkHandleInvariants,
      BufferList(..),
      HandleType(..),
      isReadableHandleType, isWritableHandleType, isReadWriteHandleType,
      isAppendHandleType,
      BufferMode(..),
      BufferCodec(..),
      NewlineMode(..), Newline(..), nativeNewline,
      universalNewlineMode, noNewlineTranslation, nativeNewlineMode
  ) where

import GHC.Internal.IO.Handle.Types
