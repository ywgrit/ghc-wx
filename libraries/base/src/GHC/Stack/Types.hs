{-# LANGUAGE Safe #-}
{-# OPTIONS_HADDOCK not-home #-}

-- |
--
-- Module      :  GHC.Stack.Types
-- Copyright   :  (c) The University of Glasgow 2015
-- License     :  see libraries/ghc-prim/LICENSE
--
-- Maintainer  :  ghc-devs@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC Extensions)
--
-- Type definitions for implicit call-stacks.
-- Use "GHC.Stack" from the base package instead of importing this
-- module directly.
--
-- /The API of this module is unstable and not meant to be consumed by the general public./
-- If you absolutely must depend on it, make sure to use a tight upper
-- bound, e.g., @base < 4.X@ rather than @base < 5@, because the interface can
-- change rapidly without much warning.
--

module GHC.Stack.Types
    (-- *  Implicit call stacks
     CallStack(..),
     HasCallStack,
     emptyCallStack,
     freezeCallStack,
     fromCallSiteList,
     getCallStack,
     pushCallStack,
     -- *  Source locations
     SrcLoc(..)
     ) where

import GHC.Internal.Stack.Types
