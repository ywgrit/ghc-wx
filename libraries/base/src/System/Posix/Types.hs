-- |
--
-- Module      :  System.Posix.Types
-- Copyright   :  (c) The University of Glasgow 2002
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  provisional
-- Portability :  non-portable (requires POSIX)
--
-- POSIX data types: Haskell equivalents of the types defined by the
-- @\<sys\/types.h>@ C header on a POSIX system.
--

module System.Posix.Types
    (-- *  POSIX data types
     -- **  Platform differences
     -- |  This module contains platform specific information about types.
     -- __/As such the types presented on this page reflect the platform
     -- on which the documentation was generated and may not coincide with
     -- the types on your platform./__
     module GHC.Internal.System.Posix.Types
     ) where

import GHC.Internal.System.Posix.Types
