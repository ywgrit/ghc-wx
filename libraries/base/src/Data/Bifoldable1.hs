-- |
-- Copyright: Edward Kmett, Oleg Grenrus
-- License: BSD-3-Clause
--

{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE Safe              #-}

module Data.Bifoldable1 where

import Control.Applicative (Const (..))
import Data.Bifoldable     (Bifoldable (..))
import Data.Semigroup      (Arg (..), Semigroup (..))
import Prelude             (Either (..), id)

class Bifoldable t => Bifoldable1 t where
     bifold1 :: Semigroup m => t m m -> m
     bifold1 = bifoldMap1 id id
     {-# INLINE bifold1 #-}

     bifoldMap1 :: Semigroup m => (a -> m) -> (b -> m) -> t a b -> m

instance Bifoldable1 Arg where
    bifoldMap1 f g (Arg a b) = f a <> g b

instance Bifoldable1 Either where
    bifoldMap1 f _ (Left a) = f a
    bifoldMap1 _ g (Right b) = g b
    {-# INLINE bifoldMap1 #-}

instance Bifoldable1 (,) where
    bifoldMap1 f g (a, b) = f a <> g b
    {-# INLINE bifoldMap1 #-}

instance Bifoldable1 ((,,) x) where
    bifoldMap1 f g (_,a,b) = f a <> g b
    {-# INLINE bifoldMap1 #-}

instance Bifoldable1 ((,,,) x y) where
    bifoldMap1 f g (_,_,a,b) = f a <> g b
    {-# INLINE bifoldMap1 #-}

instance Bifoldable1 ((,,,,) x y z) where
    bifoldMap1 f g (_,_,_,a,b) = f a <> g b
    {-# INLINE bifoldMap1 #-}

instance Bifoldable1 Const where
    bifoldMap1 f _ (Const a) = f a
    {-# INLINE bifoldMap1 #-}
