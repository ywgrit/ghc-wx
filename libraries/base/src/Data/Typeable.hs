{-# LANGUAGE Safe #-}

-- |
--
-- Module      :  Data.Typeable
-- Copyright   :  (c) The University of Glasgow, CWI 2001--2004
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  stable
-- Portability :  portable
--
-- The 'Typeable' class reifies types to some extent by associating type
-- representations to types. These type representations can be compared,
-- and one can in turn define a type-safe cast operation. To this end,
-- an unsafe cast is guarded by a test for type (representation)
-- equivalence. The module "Data.Dynamic" uses Typeable for an
-- implementation of dynamics. The module "Data.Data" uses Typeable
-- and type-safe cast (but not dynamics) to support the \"Scrap your
-- boilerplate\" style of generic programming.
--
-- == Compatibility Notes
--
-- Since GHC 8.2, GHC has supported type-indexed type representations.
-- "Data.Typeable" provides type representations which are qualified over this
-- index, providing an interface very similar to the "Typeable" notion seen in
-- previous releases. For the type-indexed interface, see "Type.Reflection".
--
-- Since GHC 7.10, all types automatically have 'Typeable' instances derived.
-- This is in contrast to previous releases where 'Typeable' had to be
-- explicitly derived using the @DeriveDataTypeable@ language extension.
--
-- Since GHC 7.8, 'Typeable' is poly-kinded. The changes required for this might
-- break some old programs involving 'Typeable'. More details on this, including
-- how to fix your code, can be found on the
-- <https://gitlab.haskell.org/ghc/ghc/wikis/ghc-kinds/poly-typeable PolyTypeable wiki page>
--

module Data.Typeable
    (-- *  The Typeable class
     Typeable,
     typeOf,
     typeRep,
     -- *  Propositional equality
     (:~:)(Refl),
     (:~~:)(HRefl),
     -- *  Type-safe cast
     cast,
     eqT,
     heqT,
     decT,
     hdecT,
     gcast,
     -- *  Generalized casts for higher-order kinds
     gcast1,
     gcast2,
     -- *  A canonical proxy type
     Proxy(..),
     -- *  Type representations
     TypeRep,
     rnfTypeRep,
     showsTypeRep,
     mkFunTy,
     -- *  Observing type representations
     funResultTy,
     splitTyConApp,
     typeRepArgs,
     typeRepTyCon,
     typeRepFingerprint,
     -- *  Type constructors
     TyCon,
     tyConPackage,
     tyConModule,
     tyConName,
     rnfTyCon,
     tyConFingerprint,
     -- *  For backwards compatibility
     typeOf1,
     typeOf2,
     typeOf3,
     typeOf4,
     typeOf5,
     typeOf6,
     typeOf7,
     trLiftedRep
     ) where

import GHC.Internal.Data.Typeable