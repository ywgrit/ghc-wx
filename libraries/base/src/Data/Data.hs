{-# LANGUAGE Safe #-}

-- |
-- Module      :  Data.Data
-- Copyright   :  (c) The University of Glasgow, CWI 2001--2004
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  stable
-- Portability :  non-portable (local universal quantification)
--
-- This module provides the 'Data' class with its primitives for
-- generic programming, along with instances for many datatypes. It
-- corresponds to a merge between the previous "Data.Generics.Basics"
-- and almost all of "Data.Generics.Instances". The instances that are
-- not present in this module were moved to the
-- @Data.Generics.Instances@ module in the @syb@ package.
--
-- \"Scrap your boilerplate\" --- Generic programming in Haskell.  See
-- <https://wiki.haskell.org/Research_papers/Generics#Scrap_your_boilerplate.21>.
--

module Data.Data (

        -- * Module Data.Typeable re-exported for convenience
        module Data.Typeable,

        -- * The Data class for processing constructor applications
        Data(
                gfoldl,
                gunfold,
                toConstr,
                dataTypeOf,
                dataCast1,      -- mediate types and unary type constructors
                dataCast2,      -- mediate types and binary type constructors
                -- Generic maps defined in terms of gfoldl
                gmapT,
                gmapQ,
                gmapQl,
                gmapQr,
                gmapQi,
                gmapM,
                gmapMp,
                gmapMo
            ),

        -- * Datatype representations
        DataType,       -- abstract
        -- ** Constructors
        mkDataType,
        mkIntType,
        mkFloatType,
        mkCharType,
        mkNoRepType,
        -- ** Observers
        dataTypeName,
        DataRep(..),
        dataTypeRep,
        -- ** Convenience functions
        repConstr,
        isAlgType,
        dataTypeConstrs,
        indexConstr,
        maxConstrIndex,
        isNorepType,

        -- * Data constructor representations
        Constr,         -- abstract
        ConIndex,       -- alias for Int, start at 1
        Fixity(..),
        -- ** Constructors
        mkConstr,
        mkConstrTag,
        mkIntegralConstr,
        mkRealConstr,
        mkCharConstr,
        -- ** Observers
        constrType,
        ConstrRep(..),
        constrRep,
        constrFields,
        constrFixity,
        -- ** Convenience function: algebraic data types
        constrIndex,
        -- ** From strings to constructors and vice versa: all data types
        showConstr,
        readConstr,

        -- * Convenience functions: take type constructors apart
        tyconUQname,
        tyconModule,

        -- * Generic operations defined in terms of 'gunfold'
        fromConstr,
        fromConstrB,
        fromConstrM

  ) where

import GHC.Internal.Data.Data
import Data.Typeable
