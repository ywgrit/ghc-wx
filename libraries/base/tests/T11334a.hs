{-# LANGUAGE Haskell2010 #-}
{-# LANGUAGE DataKinds #-}
module Main (main) where

import Data.Typeable
import GHC.Types

main :: IO ()
main = do
  print (typeOf (Proxy :: Proxy 'Just))
  print (typeOf (Proxy :: Proxy (TYPE ('BoxedRep 'Lifted))))
  print (typeOf (Proxy :: Proxy (TYPE ('BoxedRep 'Unlifted))))
