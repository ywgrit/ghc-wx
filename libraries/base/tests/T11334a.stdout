Proxy (* -> Maybe *) ('Just *)
Proxy * *
Proxy * (TYPE ('BoxedRep 'Unlifted))
